File: python-codes/msp.py
class MSP:
    STATE_IDLE = 0
    STATE_HEADER_START = 1
    STATE_HEADER_M = 2
    STATE_HEADER_ARROW = 3
    STATE_HEADER_SIZE_1 = 4
    STATE_HEADER_SIZE_2 = 5
    STATE_HEADER_CMD = 6

    REPLY_NO = 0
    REPLY_OK = 1
    REPLY_NG = 2

    def __init__(self, write, callback):
        self._packet_size = 2048
        self._state = MSP.STATE_IDLE
        self._cmd = 0
        self._size = 0
        self._packet = bytearray(self._packet_size)
        self._pos = 0

        self._seq = 0
        self._sub_seq = 0
        self._write = write
        self._cb = callback

    def send(self, cmd, data, size, reply=REPLY_NO, isSeqHeader=False):
        reply_tbl = ['>', '<', '!']
        szPacket = 0
        szHeader = 0
        pos = 0

        szPacket = (self._packet_size - 2) if isSeqHeader else self._packet_size
        szHeader = 8 if isSeqHeader else 6
        header = []

        while size >= 0:
            frag = szPacket if size > szPacket else size
            sz_ext = (frag + 2) if isSeqHeader else frag
            header.append(ord('$'))
            header.append(ord('M'))
            header.append(ord(reply_tbl[reply]))
            header.append((sz_ext >> 8) & 0xff)
            header.append(sz_ext & 0xff)
            header.append(cmd)

            if frag > 0:
                size -= frag

            if isSeqHeader:
                header.append(self._seq)
                ss = (0x80 | self._sub_seq) if size == 0 else self._sub_seq
                header.append(ss)
                self._sub_seq += 1
            self._write(bytes(header))

            if frag > 0:
                buf = data[pos:pos + frag]
                self._write(buf)
                pos += frag

            if size == 0:
                break

        if isSeqHeader:
            self._seq += 1
            self._sub_seq = 0

    def callback(self, cmd, data):
        if self._cb is not None:
            resp = self._cb(cmd, data)
            if resp is not None:
                self.send(cmd, resp, len(resp), MSP.REPLY_OK)

    def processRx(self, buf):
        for ch in buf:
            # print(
            #     f'{ch:02X} {self._state:1d} {self._pos:3d} / {self._size:3d}')
            if self._state == MSP.STATE_IDLE:
                if ch == ord('$'):
                    self._state = MSP.STATE_HEADER_START

            elif self._state == MSP.STATE_HEADER_START:
                self._state = MSP.STATE_HEADER_M if ch == ord(
                    'M') else MSP.STATE_IDLE

            elif self._state == MSP.STATE_HEADER_M:
                self._state = MSP.STATE_HEADER_ARROW if ch == ord(
                    '>') else MSP.STATE_IDLE

            elif self._state == MSP.STATE_HEADER_ARROW:
                self._size = ch << 8
                self._state = MSP.STATE_HEADER_SIZE_1

            elif self._state == MSP.STATE_HEADER_SIZE_1:
                self._size |= ch
                if self._size > self._packet_size:
                    self._state = MSP.STATE_IDLE
                    continue
                self._pos = 0
                self._state = MSP.STATE_HEADER_SIZE_2

            elif self._state == MSP.STATE_HEADER_SIZE_2:
                self._cmd = ch
                self._state = MSP.STATE_HEADER_CMD
                if self._size == 0:
                    self.callback(self._cmd, self._packet)
                    self._state = MSP.STATE_IDLE

            elif self._state == MSP.STATE_HEADER_CMD:
                if self._pos < self._size:
                    self._packet[self._pos] = ch
                    self._pos = self._pos + 1
                    # self._packet.append(ch)
                if self._pos == self._size:
                    self.callback(self._cmd, self._packet)
                    self._state = MSP.STATE_IDLE

File: rover-codes/src/MSP.cpp
#include "MSP.h"

MSP::MSP(uint16_t wPacketSize) {
    m_stateRx    = STATE_IDLE;
    m_wOffset   = 0;
    m_wDataSize = 0;
    m_ucSeq      = 0;
    m_ucSubSeq   = 0;
    m_wPacketSize = wPacketSize;
    m_pRxPackets = new uint8_t[m_wPacketSize];
    m_pResultPackets = new uint8_t[m_wPacketSize];
}

MSP::~MSP() {
    delete m_pRxPackets;
    delete m_pResultPackets;
}

void MSP::setInterface(MSPInterface *pInterface) {
    m_pInterface = pInterface;
}

void MSP::setCallback(MSPCallback *pCallback) {
    m_pCallback  = pCallback;
}

void MSP::send(uint8_t cmd, uint8_t *data, uint16_t size, reply_t reply, bool isSeqHeader) {
    const uint8_t reply_tbl[3] = { '>', '<', '!'};
    uint8_t  header[8];
    uint8_t *buf = data;
    uint16_t szPacket;
    uint16_t szHeader;

    if (m_pInterface) {
        szPacket = isSeqHeader ? (m_wPacketSize - 2) : m_wPacketSize;
        szHeader = isSeqHeader ? 8 : 6;

        while (size >= 0) {
            uint16_t frag   = (size > szPacket) ? szPacket : size;
            uint16_t sz_ext = isSeqHeader ? frag + 2 : frag;

            header[0] = '$';
            header[1] = 'M';
            header[2] = reply_tbl[reply];
            header[3] = (sz_ext >> 8) & 0xff;
            header[4] = sz_ext & 0xff;
            header[5] = cmd;

            if (frag > 0)
                size -= frag;

            // 2bytes additional header = should be considered as data
            if (isSeqHeader) {
                header[6] = m_ucSeq;
                header[7] = (size == 0) ? (0x80 | m_ucSubSeq) : (m_ucSubSeq);
                m_ucSubSeq++;
            }
            m_pInterface->write(header, szHeader);

            if (frag > 0) {
                m_pInterface->write(buf, frag);
                buf  += frag;
            }
            if (size == 0)
                break;
        }

        if (isSeqHeader) {
            m_ucSeq++;
            m_ucSubSeq = 0;
        }
    }
}

void MSP::callback(uint8_t cmd, uint8_t *pData, uint16_t size, uint8_t *pRes) {
    if (m_pCallback) {
        int16_t ret = m_pCallback->onCommand(cmd, pData, size, pRes);
        if (ret >= 0) {
            send(cmd, pRes, ret, REPLY_OK);
        }
    }
}

void MSP::processRx(uint8_t *pBuf, uint16_t size) {
    for (uint16_t i = 0; i < size; i++) {
        uint8_t ch = pBuf[i];

        switch (m_stateRx) {
            case STATE_IDLE:
                if (ch == '$')
                    m_stateRx = STATE_HEADER_START;
                break;

            case STATE_HEADER_START:
                m_stateRx = (ch == 'M') ? STATE_HEADER_M : STATE_IDLE;
                break;

            case STATE_HEADER_M:
                m_stateRx = (ch == '>') ? STATE_HEADER_ARROW : STATE_IDLE;
                break;

            case STATE_HEADER_ARROW:
                m_wDataSize = (ch << 8);
                m_stateRx    = STATE_HEADER_SIZE_1;
                break;

            case STATE_HEADER_SIZE_1:
                m_wDataSize |= ch;
                if (m_wDataSize > m_wPacketSize) {
                    m_stateRx = STATE_IDLE;
                    continue;
                }
                m_wOffset = 0;
                m_stateRx  = STATE_HEADER_SIZE_2;
                break;

            case STATE_HEADER_SIZE_2:
                m_ucCmd    = ch;
                m_stateRx  = STATE_HEADER_CMD;
                if (m_wDataSize == 0) {
                    callback(m_ucCmd, m_pRxPackets, m_wDataSize, m_pResultPackets);
                    m_stateRx = STATE_IDLE;
                }
                break;

            case STATE_HEADER_CMD:
                if (m_wOffset < m_wDataSize) {
                    m_pRxPackets[m_wOffset++] = ch;
                }

                if (m_wOffset == m_wDataSize) {
                    callback(m_ucCmd, m_pRxPackets, m_wDataSize, m_pResultPackets);
                    m_stateRx = STATE_IDLE;
                }
                break;
        }
    }
}
